<a href="https://colab.research.google.com/github/sahala-mol/ds-ml/blob/main/sahala_mol_vs__KNN.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#### Problem Statement

Nowadays, social media advertising is one of the popular forms of advertising. Advertisers can utilise user's demographic information and target their ads accordingly.  You are given a dataset having the following attributes:

|Field|Description|
|---:|:---|
|UserID|Unique ID|
|Gender|Male or Female|
|Age|Age of a person|
|EstimatedSalary|Salary of a person|
|Purchased|‘0’ or ‘1’. ‘0’ means not purchased and ‘1’ means purchased.|


**Source:** https://www.kaggle.com/rishabhsingh98/social-network-ads

**Citation:** Rishabh Singh. (2020). Social Network Ads.

Implement kNN Classifier to determine whether a user will purchase a particular product displayed on a social network ad or not.

---

### List of Activities

**Activity 1:** Import Modules and Read Data

  
**Activity 2:**  Perform Train-Test Split

**Activity 3:**  Determine the Optimal Value of  $k$

**Activity 4:** Build kNN Classifier Model






---

#### Activity 1: Import Modules and Read Data

Import the necessary Python packages.

Read the data from a CSV file to create a Pandas DataFrame.

**Dataset-->**  social-network-ads.csv

Also, print the first five rows of the dataset. Check for null values and treat them accordingly.


In [2]:
# Import all the necessary packages
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
# Load the dataset
df=pd.read_csv('/content/social-network-ads - social-network-ads.csv')
# Print first five rows using head() function
df.head()

User ID  Gender  Age  EstimatedSalary  Purchased
0  15624510    Male   19            19000          0
1  15810944    Male   35            20000          0
2  15668575  Female   26            43000          0
3  15603246  Female   27            57000          0
4  15804002    Male   19            76000          0

In [3]:
# Check if there are any null values. If any column has null values, treat them accordingly
df.isnull().sum()

User ID            0
Gender             0
Age                0
EstimatedSalary    0
Purchased          0
dtype: int64

**Q:** Are there any missing or null values in the dataset?

**A:** No.

---

#### Activity 2: Perform Train-Test Split

In this dataset, `Purchased` is the target variable and all other columns other than `Purchased` are feature variables.

Create two separate DataFrames, one containing the feature variables and the other containing the target variable. Also, drop the `User ID` column from the features DataFrame as it is of no use.





In [4]:
# Split the dataset into dependent and independent features
features_df=df.drop(['Purchased','User ID'],axis=1)
target_df=df['Purchased']
features_df.head()
target_df.head()

0    0
1    0
2    0
3    0
4    0
Name: Purchased, dtype: int64

Print the summary of features DataFrame to determine the data type of each feature variable.

In [5]:
# Use 'info()' function with the features DataFrame.
features_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Gender           400 non-null    object
 1   Age              400 non-null    int64 
 2   EstimatedSalary  400 non-null    int64 
dtypes: int64(2), object(1)
memory usage: 9.5+ KB


Convert categorical `Gender` feature into numerical  by calling the `get_dummies()` function of `pandas` module and passing features DataFrame as input.




In [6]:
# Use 'get_dummies()' function to convert each categorical column in a DataFrame to numerical.
x=pd.get_dummies(features_df)
x.head(5)

Age  EstimatedSalary  Gender_Female  Gender_Male
0   19            19000          False         True
1   35            20000          False         True
2   26            43000           True        False
3   27            57000           True        False
4   19            76000          False         True

Split the dataset into train set and test set such that the train set contains 70% of the instances and the remaining instances will become the test set.

In [7]:
# Split the DataFrame into the train and test sets.
# Perform train-test split using 'train_test_split' function.
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,target_df,test_size=0.3,random_state=42)
# Print the shape of the train and test sets.
print(x_train.shape)
print(x_test.shape)


(280, 4)
(120, 4)


After this activity, you must obtain train and test sets so that they can be used for training and testing the kNN Classifier.

#### Activity 4: Build kNN Classifier Model

Deploy the kNN Classifier model for the optimal value of $k$ using the steps given below:   

1. Import the `KNeighborsClassifier` class from the `sklearn.neighbors` module (if not imported yet).

2. Create an object of `KNeighborsClassifier` and pass the optimal $k$ value as 5 to its constructor.

3. Call the `fit()` function using the classifier object and pass the train set as inputs to this function.

4. Perform prediction for train and test sets using the `predict()` function.

5. Also, determine the accuracy score of the train and test sets using the `score()` function.

In [14]:
# Train kNN Classifier model
from sklearn.neighbors import KNeighborsClassifier
knn=KNeighborsClassifier(n_neighbors=3)
knn.fit(x_train,y_train)
# Perform prediction using 'predict()' function.
y_train_pred=knn.predict(x_train)
y_test_pred=knn.predict(x_test)
# Call the 'score()' function to check the accuracy score of the train set and test set.
from sklearn.metrics import accuracy_score
print(accuracy_score(y_train,y_train_pred))
print(accuracy_score(y_test,y_test_pred))

0.8892857142857142
0.7916666666666666


Print the classification report to get an in-depth overview of the classifier performance using the `classification_report()` function of `sklearn.metrics` module.

In [15]:
# Display the precision, recall, and f1-score values.
from sklearn.metrics import classification_report
print(classification_report(y_test,y_test_pred))

              precision    recall  f1-score   support

           0       0.79      0.89      0.84        73
           1       0.79      0.64      0.71        47

    accuracy                           0.79       120
   macro avg       0.79      0.76      0.77       120
weighted avg       0.79      0.79      0.79       120



**Q:** Write down the f1-scores for both the target labels.

**A:** 0.84 & 0.71




---

